### PR TITLE
Pufferbomb

### PR DESCRIPTION
**Pull request by jocelynzaman**
PufferbombSpawner is a trigger for moving a group of Pufferbombs up at a given height and speed.
- Pufferbombs need to be children of the PufferbombSpawner.
 -Pufferbombs will have their script initially disabled until the trigger collision is detected.

Pufferbombs is a stationary enemy that will explode and may deal damage to the Player.
 - Pufferbombs will remain stationary after the initial float up from the PufferbombSpawner. 
 - When the Player is at a certain distance from the Pufferbomb, the Pufferbomb will explode and deal damage to the Player. 
 - The Player can also shoot the Pufferbomb and the Pufferbomb will explode.

### CODE REVIEW

File: Assets/Prefabs/Enemies/Pufferbomb.prefab [hunk 1]
```diff
@@ -0,0 +1,4952 @@
+%YAML 1.1
```
Comment (nicholasngoc) [inline Assets/Prefabs/Enemies/Pufferbomb.prefab:1]: Plz zero out/reset the transform

File: Assets/Prefabs/Enemies/PufferbombSpawner.prefab [hunk 2]
```diff
@@ -0,0 +1,61 @@
+%YAML 1.1
```
Comment (nicholasngoc) [inline Assets/Prefabs/Enemies/PufferbombSpawner.prefab:1]: Plz zero out/reset transform

File: Assets/Scripts/Enemies/Pufferbomb.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        explosion = transform.GetComponent<ParticleSystem>();
+        mesh = transform.GetComponent<MeshRenderer>();
+
+        isExploding = false;
+        isShot = false;
+    }
+
+    void Update()
+    {
+        if (!isExploding && !isShot && Vector3.Distance(PlayerInfo.singleton.transform.position, transform.position) <= distance)
```
Comment (nicholasngoc) [inline Assets/Scripts/Enemies/Pufferbomb.cs:26]: It's not very good to calculate the distance within the update function. I would add a new sphere collider on the pufferbomb prefab that's a trigger and then you can use OnTriggerEnter() to determine when the player is close enough for it to explode. Also i'm unsure why but when testing the puffer fish was not exploding when I ran into it

File: Assets/Scripts/Enemies/Pufferbomb.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    }
+
+    //Play explosion particle system effect
+    private IEnumerator Explode()
+    {
+        isExploding = true;
+        mesh.enabled = false;
+        explosion.Play(); //start Particle System explosion effect
+        yield return new WaitForSeconds(1.2f); //wait for explosion to finish before destroying Pufferbomb
```
Comment (nicholasngoc) [inline Assets/Scripts/Enemies/Pufferbomb.cs:39]: Please turn the 1.2f into an instance variable. It'll help designers if they need to adjust this value

File: Assets/Scripts/Enemies/PufferbombSpawner.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    private float finalHeight;
+
+    void Start()
+    {
+        //disable Pufferbomb script for all Pufferbombs
+        foreach (Transform child in transform)
+        {
+            if (child.GetComponent<Pufferbomb>() != null)
+            {
+                child.GetComponent<Pufferbomb>().enabled = true;
```
Comment (nicholasngoc) [inline Assets/Scripts/Enemies/PufferbombSpawner.cs:20]: I'm a little confused here. At first I read that you disabled them until the trigger was activated but then you activated here? Looking into it further I am unsure if this is needed. Please justify why this is needed and if not then please remove

**Review by nicholasngoc — CHANGES_REQUESTED** [review]
Almost there

**Comment by jocelynzaman** [comment]
Thanks for the feedback. I'll try to resolve these issues before Friday's meeting.

File: Assets/Scripts/Enemies/PufferbombSpawner.cs (hunk 5 above)
Comment (jocelynzaman) [reply Assets/Scripts/Enemies/PufferbombSpawner.cs:20]: I think I might have copied and pasted without changing it to false, so I fixed it.

File: Assets/Scripts/Enemies/Pufferbomb.cs (hunk 4 above)
Comment (jocelynzaman) [reply Assets/Scripts/Enemies/Pufferbomb.cs:39]: I added a public explosionTime variable.

File: Assets/Scripts/Enemies/Pufferbomb.cs (hunk 3 above)
Comment (jocelynzaman) [reply Assets/Scripts/Enemies/Pufferbomb.cs:26]: I added the sphere collider trigger. I'm not sure if my implementation is efficient though, so would like to know if it needs to be revised.

**Review by propredator23 — APPROVED** [review]
Looks Good

**Pull request merged**